### PR TITLE
README: revamp orienting it to usage and extract development readme

### PR DESCRIPTION
**Pull request by miguelsorianod**
Development/Contribution oriented documentation has been moved to the file DEVELOPMENT.md. References to 3scale_core have been updated to the new upstream name 'pisoni'.

README has been rewritten orienting it to use the Apisonator product and a general description of it.

Still pending to fix the issue with broken links to documentation in the old 3scale website.

Related to issue https://github.com/3scale/apisonator/issues/27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,227 +4,178 @@
...
+## Development
 
-Next cd into the directory, `cd apisonator`.
-
-### Containerized environment
-
-#### Prerequisites
-
-* Docker (requires version 1.10.0 or later)
+See the file ```DEVELOPMENT.md```
```
Comment (eguzki) [inline README.md]: better a link
```
[DEVELOPMENT](DEVELOPMENT.md)
```

Comment (miguelsorianod) [reply README.md]: :+1:

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+This command will take care of downloading and building all dependencies. Once
+that is done, the process will be way faster the next time.
+
+The project's source code will be available in `~/apisonator` and sync'ed with
+your local apisonator directory, so you can edit files in your preferred
+environment and still be able to run whatever you need inside the Docker
+container.
+
+The listener service port (3000 by default) is automatically forwarded to the host machine. The `make dev` command can be executed with the environment variable `PORT` set to
```
Comment (unleashed) [inline DEVELOPMENT.md]: really long lines here, hard to edit - do you have a way to autowrap long lines so that a newline char is inserted when too long?

File: README.md [hunk 3]
```diff
@@ -4,227 +4,178 @@
...
+## Development
 
-Next cd into the directory, `cd apisonator`.
-
-### Containerized environment
-
-#### Prerequisites
-
-* Docker (requires version 1.10.0 or later)
+See the file [DEVELOPMENT](DEVELOPMENT.md)
```
Comment (unleashed) [inline README.md]: maybe move this section way down so that when reading this for the first time you get straight to the explanation of what is this project

File: README.md [hunk 4]
```diff
@@ -4,227 +4,178 @@
...
 
-Changes in code sometimes translate into changes in dependencies. If that is the
-case, your container will lag behind in dependencies, and some things might just
-start breaking. You might want to make a habit of making sure dependencies are
-updated when you enter your container. Run this from the project directory:
+Three main operations can be performed with this API:
 
-> $ `bundle install`
+ * Report: Reports a number of hits to one or more metrics
+ * Auth: Authorize a request
```
Comment (unleashed) [inline README.md]: This should be `Authorize`.

File: README.md [hunk 5]
```diff
@@ -4,227 +4,178 @@
...
+ * Authrep: Combination of both the above
 
-The container image has additional tools to handle these dependencies for
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+Make sure to read the corresponding documentation in [ActiveDocs](https://support.3scale.net/reference/active-docs) for
```
Comment (unleashed) [inline README.md]: the link is broken - we should ask for the correct link

File: README.md (hunk 4 above)
Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md (hunk 5 above)
Comment (miguelsorianod) [reply README.md]: Yes, I was looking into that as I said on the initial comment of the PR :+1: I'll update the PR with the fix once I have the new link (if available)

File: README.md [hunk 6]
```diff
@@ -4,227 +4,178 @@
...
+ * Perform authorization of the requests
+This two actions can be configured to be performed in different Redis databases if desired (see the Configuration section).
 
-#### Workflow
+Finally, another API is provided to configure services in Apisonator. This API is named
+'Internal API'. You can also generate the documentation of the Internal API with Rake tasks.
 
-Your project directory within the container is sync'ed with your local clone of
-the project so that changes in one reflect instantly in the other.
+For a quick Apisonator test, Random services can be created and configured in backend via
```
Comment (unleashed) [inline README.md]: backend => apisonator

Comment (unleashed) [reply README.md]: perhaps even rephrase it to not have to mention apisonator twice

File: README.md [hunk 7]
```diff
@@ -4,227 +4,194 @@
...
-## Running tests
+### Apisonator failed jobs rescheduler
 
-You can either run them manually in the container-based development environment
-or have a container launched just for running the tests. For the latter you just
-need to run `make test`, and you can configure any additional environment
-variables for the test scripts like so:
+This is a simple task that acts as a cron scheduler to requeue jobs that failed
+when being processed by an Apisonator worker. The jobs are requeued into
+the Redis database.
```
Comment (unleashed) [inline README.md]: I think it is best to refer to the background jobs queue database (we can mention Resque using Redis when we define it above).

File: DEVELOPMENT.md (hunk 2 above)
Comment (miguelsorianod) [reply DEVELOPMENT.md]: I have updated the README and DEVELOPMENT documents to try to stick the lines more or less at 80 characters. In my IDE I have a vertical line as a guide to this

File: README.md [hunk 8]
```diff
@@ -4,227 +4,178 @@
...
-a different desired listener port value to forward on the host machine.
+It has the following components:
 
-This Docker container is persistent, so your changes will be kept the next time
-you enter it.
+### Apisonator listener
 
-Getting rid of the persistent container is done with `make dev-clean`, whereas
-removing its image is done using `make dev-clean-image`.
+It provides the point of entry of the API Management Platform's Backend. The Service management API (SM API) is provided to authorize and report consumer
```
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md [hunk 9]
```diff
@@ -4,227 +4,178 @@
...
 
-Changes in code sometimes translate into changes in dependencies. If that is the
-case, your container will lag behind in dependencies, and some things might just
-start breaking. You might want to make a habit of making sure dependencies are
-updated when you enter your container. Run this from the project directory:
+Make sure to read the corresponding documentation in [ActiveDocs](https://support.3scale.net/reference/active-docs) for
+more information.
 
-> $ `bundle install`
+It attempts to respond with lowest possible latency, performing minimal work and offloading other work to backend-workers by enqueuing
```
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply README.md]: Thank you. I have renamed the non-code references where 'backend' appeared to 'Apisonator'

File: README.md (hunk 6 above)
Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md [hunk 10]
```diff
@@ -4,227 +4,194 @@
...
+Docker or by placing them in a ENV file and setting the ENV file in Docker via
+the `--env-file` flag.
 
-where in this last case the [ ] characters must be literally placed.
+The most important variables to set are:
 
-Finally, you can also execute tests with rspec directly with:
+ * CONFIG_QUEUES_MASTER_NAME: Set this to <hostname>:<port> of where the Redis
+   has been installed
+ * CONFIG_REDIS_PROXY: Set this to <hostname>:<port> of where the Redis
```
Comment (unleashed) [inline README.md]: This and the other setting one above receive a Redis URL - hostname:port is deprecated and discouraged. We might want to mention support for Sentinels (in which case these two settings change their meaning) but that is complex enough that might be better explained in its own section.

File: README.md [hunk 11]
```diff
@@ -4,227 +4,194 @@
...
+docker run -it amp:apisonator-<version_number> -p 3001:3001 --env-file <myenv_file> 3scale_backend start -p 3001 -x /dev/stdout
+```
 
-Make sure to read the corresponding documentation in [ActiveDocs](https://support.3scale.net/reference/active-docs) for the external
-API. You can also generate the documentation of the Internal API with Rake tasks.
+You can see all the available options of the apisonator listener by executing:
 
-## Integration flow
+```
+docker run -it amp:apisonator-<version_number> 3scale_backend --help
```
Comment (unleashed) [inline README.md]: just `help` is a better form, as it is a command that can receive other commands as argument.

Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md (hunk 7 above)
Comment (miguelsorianod) [reply README.md]: :+1:

**Comment by miguelsorianod** [comment]
I have replaced the broken ActiveDocs link with a reference to the Swagger-generated documentation that we have available in 'docs/active_docs/Service Management API.json' because the link is not available anymore and there is no global place to look at it (access to the 3scale admin portal is needed)

File: README.md [hunk 12]
```diff
@@ -4,227 +4,198 @@
...
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+This component needs access to a Redis database to perform the following actions:
+ * Enqueue reports, which will be processed by the Apisonator worker component
+ * Perform authorization of the requests
+This two actions can be configured to be performed in different Redis
+databases if desired (see the Prerequisites and Configuration sections).
 
-#### Workflow
+Finally, another API is provided to configure services in Apisonator. This API
+is named 'Internal API'. You can also generate the documentation of the
```
Comment (unleashed) [inline README.md]: let's make it clear here that this API is not meant for general consumption but for administrative operations, so it must be protected and/or not exposed to untrusted parties.

Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md [hunk 13]
```diff
@@ -4,227 +4,200 @@
...
-> $ `bundle install`
+It attempts to respond with lowest possible latency, performing minimal work
+and offloading other work to Apisonator workers by enqueuing tasks into job queues.
 
-The container image has additional tools to handle these dependencies for
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+This component needs access to a Redis database to perform the following actions:
+ * Enqueue reports, which will be processed by the Apisonator worker component
+ * Perform authorization of the requests
+This two actions can be configured to be performed in different Redis
```
Comment (unleashed) [inline README.md]: typo `These`

File: README.md [hunk 14]
```diff
@@ -4,227 +4,200 @@
...
+in Apisonator. This API is intended only for administrative purposes and not
+for general consumption. Therefore, usage of this API should be protected or
+not exposed to untrusted parties. You can also generate the documentation of
+the Internal API with Rake tasks.
 
-Your project directory within the container is sync'ed with your local clone of
-the project so that changes in one reflect instantly in the other.
+To quickly test Apisonator, random services can be created and configured on it
+via the use of the 'Buddhi' tool located in our performance tests
+toolkit: [3scale perftest-toolkit](https://github.com/3scale/perftest-toolkit/).
```
Comment (unleashed) [inline README.md]: also mention the pisoni Ruby client for the internal API

File: README.md [hunk 15]
```diff
@@ -4,227 +4,200 @@
...
 
-We recommend editing code and committing locally, and executing tests within the
-container, since the container won't have your own tools and configurations.
+This component may also be referred as '3scale_backend'.
```
Comment (unleashed) [inline README.md]: referred _to_ as '3scale_backend'

File: README.md [hunk 16]
```diff
@@ -4,227 +4,200 @@
...
 
-### On local machine (unsupported)
+### Apisonator worker
 
-This is **unsupported** and **strongly discouraged** because it is the source of a lot
-of headaches and potential problems and introduces the need for us to document
-dependencies and config files here that are very likely to be outdated and
-incomplete.
+It is responsible for performing background tasks off-loaded from
+Apisonator Listeners (enqueued jobs).
```
Comment (unleashed) [inline README.md]: listener is capitalized here and probably shouldn't

File: README.md [hunk 17]
```diff
@@ -4,227 +4,200 @@
...
+This component needs access to a Redis database to perform the following actions:
+ * Dequeue the report jobs that have been submitted to the Redis database by the
+   Apisonator listener/s or the Apisonator failed jobs rescheduler
 
-```
-ThreeScale::Backend.configure do |config|
-  config.redis.proxy = 'localhost:6379'
-end
-```
+This component may also be referred as '3scale_backend_worker'.
```
Comment (unleashed) [inline README.md]: referred _to_ as...

File: README.md [hunk 18]
```diff
@@ -4,227 +4,200 @@
...
-You can either run them manually in the container-based development environment
-or have a container launched just for running the tests. For the latter you just
-need to run `make test`, and you can configure any additional environment
-variables for the test scripts like so:
+This is a simple task that acts as a cron scheduler to requeue jobs that failed
+when being processed by an Apisonator worker. The jobs are requeued into
+the Redis background jobs database.
 
-> `$ make DOCKER_OPTS="-e TEST_ALL_RUBIES=1" test`
+This component may also be referred as 'backend-cron'.
```
Comment (unleashed) [inline README.md]: same thing, referred _to_ as

File: README.md [hunk 19]
```diff
@@ -4,227 +4,200 @@
...
-> `$ script/services start`
-
-Then you can execute the following commands to execute all tests:
+* Docker (requires version 1.10.0 or later)
+* A Redis database, used to store API request statistics and services. Also
+  used to perform API requests authorizations. In Apisonator this database
+  is commonly referred as 'Redis Storage'
+* A Redis database, used to store background jobs. The Redis Resque library
+  is used for this. In Apisonator this database is commonly referred as
+  'Redis Resque', or as the 'background jobs database'
```
Comment (unleashed) [inline README.md]: maybe mention that the two databases can be combined into a single Redis instance (by using the database id specification, ie. `redis://host:port/0`, `redis://host:port/1`, etc

File: README.md [hunk 20]
```diff
@@ -4,227 +4,200 @@
...
-be found in the manifest:
+#### Apisonator Listener
 
-> `$ bundle exec 3scale_backend manifest`
+To run an Apisonator listener, the script bin/3scale_backend is used. To
+run it from a previously generated Apisonator docker image:
 
-## Documentation
+```
+docker run -it amp:apisonator-<version_number> -p 3001:3001 --env-file <myenv_file> 3scale_backend start -p 3001 -x /dev/stdout
```
Comment (unleashed) [inline README.md]: hm, we have an `EXPOSE 3000` command in the Dockerfile, maybe we can take advantage of that (plus it's also the default 3scale_backend listener port) to shorten this command here (ie. move to 3000 and we lose the -p 3001 in the start command, and maybe we can shorten the docker `-p 3000:3000` or use `--expose 3000`?

Comment (miguelsorianod) [reply README.md]: As far as I know, the -p parameter and the --expose parameter do different things. The expose parameter does not really publish the ports:
https://docs.docker.com/engine/reference/builder/
```
The EXPOSE instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published. To actually publish the port when running the container, use the -p flag on docker run to publish and map one or more ports, or the -P flag to publish all exposed ports and map them to high-order ports.
```
Regarding the port number, although in the Dockerfile it exports port 3000 I chose 3001 because the automatic Makefile rules start the backend listener in port 3001 (see file openshift/mk/backend.mk) and I decided to be consistent with that because we have the othersection with the execution via Makefile

Comment (unleashed) [reply README.md]: ok

**Comment by unleashed** [comment]
@miguelsorianod looks ok, can you squash it all into a single commit?

**Comment by unleashed** [comment]
@miguelsorianod also, is this still WIP?

**Comment by miguelsorianod** [comment]
I'm going to squash them then :+1:

**Comment by miguelsorianod** [comment]
Done. Can you review it?

Thank you.

File: README.md [hunk 21]
```diff
@@ -4,227 +4,211 @@
...
+ * Report: Reports a number of hits to one or more metrics
+ * Authorize: Authorize a request
+ * Authrep: Combination of both the above
 
-Changes in code sometimes translate into changes in dependencies. If that is the
-case, your container will lag behind in dependencies, and some things might just
-start breaking. You might want to make a habit of making sure dependencies are
-updated when you enter your container. Run this from the project directory:
+Make sure to read the corresponding Swagger-generated documentation of this
+operations, located in `docs/active_docs/Service Management API.json`
```
Comment (eguzki) [inline README.md]: Markdown link
```
[docs/active_docs/Service Management API.json](docs/active_docs/Service Management API.json)
```

File: README.md [hunk 22]
```diff
@@ -4,227 +4,211 @@
...
+It attempts to respond with lowest possible latency, performing minimal work
+and offloading other work to Apisonator workers by enqueuing tasks into job queues.
 
-The container image has additional tools to handle these dependencies for
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+This component needs access to a Redis database to perform the following actions:
+ * Enqueue reports, which will be processed by the Apisonator worker component
+ * Perform authorization of the requests
+These two actions can be configured to be performed in different Redis
+databases if desired (see the Prerequisites and Configuration sections).
```
Comment (eguzki) [inline README.md]: you can create anchor to a heading like *Prerequisites* or *Configuration*

File: README.md [hunk 23]
```diff
@@ -4,227 +4,211 @@
...
-> $ `bundle install`
+It attempts to respond with lowest possible latency, performing minimal work
+and offloading other work to Apisonator workers by enqueuing tasks into job queues.
 
-The container image has additional tools to handle these dependencies for
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+This component needs access to a Redis database to perform the following actions:
+ * Enqueue reports, which will be processed by the Apisonator worker component
+ * Perform authorization of the requests
+These two actions can be configured to be performed in different Redis
```
Comment (eguzki) [inline README.md]: Extra newline needed

File: README.md [hunk 24]
```diff
@@ -4,227 +4,211 @@
...
 
-> $ `bundle install`
+It attempts to respond with lowest possible latency, performing minimal work
+and offloading other work to Apisonator workers by enqueuing tasks into job queues.
 
-The container image has additional tools to handle these dependencies for
-multiple Ruby versions. Check out the `scripts` directory if you are curious.
+This component needs access to a Redis database to perform the following actions:
+ * Enqueue reports, which will be processed by the Apisonator worker component
+ * Perform authorization of the requests
```
Comment (eguzki) [inline README.md]: I would add a brief explanation about what *authorization* means. Something like:

```
Authorization task is based several validations. Two main validations are:
- *api key* is valid
- usage metrics are within limits
```

Feel free to change the explanation to something else for clearness


Maybe getting too much into detail. Up to you.

File: README.md (hunk 21 above)
Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md [hunk 25]
```diff
@@ -4,227 +4,211 @@
...
+The worker component takes care of running these enqueued jobs, mainly related
+to reporting of previous traffic.
 
-```
-ThreeScale::Backend.configure do |config|
-  config.redis.proxy = 'localhost:6379'
-end
-```
+This component needs access to a Redis database to perform the following actions:
+ * Dequeue the report jobs that have been submitted to the Redis database by the
```
Comment (eguzki) [inline README.md]: Maybe explanation about what to do with dequeued jobs?

File: README.md (hunk 22 above)
Comment (miguelsorianod) [reply README.md]: :+1:

File: README.md (hunk 25 above)
Comment (miguelsorianod) [reply README.md]: I'm not sure what you want to mean here. This is automatically performed by the Apisonator Workers. There is nothing "to do" with the dequeued jobs by the user.

Do you mean, "explain what is done with the dequeued jobs"?

**Comment by miguelsorianod** [comment]
I have pushed the changes related to the comments by @eguzki

File: README.md (hunk 24 above)
Comment (miguelsorianod) [reply README.md]: :+1:

**Comment by miguelsorianod** [comment]
Just to add more information, the next step in a next PR is to open a public repository with already generated Apisonator docker images and document how/where to gather this docker image and use it.
Before doing that I'd like to discuss it with @unleashed.

File: README.md [hunk 26]
```diff
@@ -4,227 +4,218 @@
...
 
-> `$ script/test`
+## Development
 
-### Running tests (advanced)
+See the file [DEVELOPMENT](DEVELOPMENT.md)
 
-In order to run the tests manually in the container-based development, the services
-needed to run them correctly must be started before with:
+## Deployment set up
```
Comment (eguzki) [inline README.md]: I would not call this section *deployment* 

This should be the tipical: "How to run".  Deployment is out of the scope of this document

File: DEVELOPMENT.md [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+> `$ script/test_external -- -d`
+
+## Documentation
+
+Make sure to read the corresponding Swagger-generated documentation,
+located in [docs/active_docs/Service Management API.json](docs%2Factive_docs%2FService%20Management%20API.json)
+for the external API.
+You can also generate the documentation of the Internal API with Rake tasks.
+
+## Integration flow
```
Comment (eguzki) [inline DEVELOPMENT.md]: Maybe better *Contributing* ??

Apicast section about contributing is very comprehensive. With PR guidelines, etc...
https://github.com/3scale/apicast#contributing

**Comment by unleashed** [comment]
@miguelsorianod what's missing at this moment in this PR?

**Comment by miguelsorianod** [comment]
For me all the pending tasks were already reviewed/done. I think @eguzki did want to take another look

**Comment by unleashed** [comment]
bors r=@unleashed,@eguzki
